File: automedal/agent/providers/anthropic.py
"""Anthropic provider — wraps the official `anthropic` SDK.

Handles three sources behind the same code path:
    base_url=None                            → Anthropic direct
    base_url="https://opencode.ai/zen/go"    → opencode-go (minimax-m2.7 etc)
    base_url=<custom>                        → any Anthropic-compatible gateway

Streaming is via `client.messages.stream(...)`. Per-token text deltas are
forwarded to the EventSink. Thinking blocks (returned by minimax-m2.7
through opencode-go) are surfaced via `events.thinking()` and echoed back
verbatim on the next turn so the upstream "thinking continuity" contract
holds.

Tool messages from our internal shape (role="tool") are repacked into
Anthropic's user-role tool_result blocks before sending.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any

from automedal.agent.providers.base import ChatProvider, ChatTurn, ToolCall, Usage


@dataclass
class AnthropicProvider:
    model: str
    api_key: str
    base_url: str | None = None
    max_tokens: int = 4096
    timeout: int = 120

    def _client(self):
        from anthropic import AsyncAnthropic
        kw: dict[str, Any] = {"api_key": self.api_key, "timeout": self.timeout}
        if self.base_url:
            kw["base_url"] = self.base_url
        return AsyncAnthropic(**kw)

    # ── outbound serialization ───────────────────────────────────────────────

    @staticmethod
    def _to_anthropic_messages(internal: list[dict]) -> list[dict]:
        """Translate our internal messages into Anthropic's wire format.

        Tool-role messages collapse into the *previous* (or next) user message
        as `tool_result` content blocks per Anthropic's API convention.
        """
        out: list[dict] = []
        pending_tool_results: list[dict] = []

        def _flush_tool_results() -> None:
            nonlocal pending_tool_results
            if pending_tool_results:
                out.append({"role": "user", "content": pending_tool_results})
                pending_tool_results = []

        for msg in internal:
            role = msg.get("role")
            if role == "tool":
                pending_tool_results.append({
                    "type": "tool_result",
                    "tool_use_id": msg["tool_use_id"],
                    "content": msg.get("content", ""),
                    **({"is_error": True} if msg.get("is_error") else {}),
                })
                continue

            # Any non-tool message ends the pending tool_result run
            _flush_tool_results()

            if role == "user":
                content = msg["content"]
                if isinstance(content, str):
                    out.append({"role": "user", "content": content})
                else:
                    out.append({"role": "user", "content": content})
            elif role == "assistant":
                out.append({"role": "assistant", "content": msg["content"]})
            else:
                raise ValueError(f"unknown message role: {role!r}")

        _flush_tool_results()
        return out

    @staticmethod
    def _tool_specs(tools: list) -> list[dict]:
        return [{"name": t.name, "description": t.description, "input_schema": t.schema} for t in tools]

    # ── chat ─────────────────────────────────────────────────────────────────

    async def chat_stream(
        self,
        *,
        system: str,
        messages: list[dict],
        tools: list,
        events: Any,
    ) -> ChatTurn:
        client = self._client()
        api_messages = self._to_anthropic_messages(messages)
        kwargs: dict[str, Any] = {
            "model": self.model,
            "max_tokens": self.max_tokens,
            "system": system,
            "messages": api_messages,
        }
        if tools:
            kwargs["tools"] = self._tool_specs(tools)

        try:
            async with client.messages.stream(**kwargs) as stream:
                async for ev in stream:
                    self._handle_stream_event(ev, events)
                final = await stream.get_final_message()
        finally:
            try:
                await client.close()
            except Exception:
                pass

        return self._final_to_turn(final)

    @staticmethod
    def _handle_stream_event(ev: Any, events: Any) -> None:
        if events is None:
            return
        # `text` events fire for each text-delta in a TextBlock
        kind = getattr(ev, "type", "")
        if kind == "text":
            txt = getattr(ev, "text", "")
            if txt:
                events.delta(txt)
        elif kind == "content_block_start":
            block = getattr(ev, "content_block", None)
            if block is not None and getattr(block, "type", "") == "tool_use":
                # Surface tool_start lazily on actual call execution; nothing here.
                pass
        # Streaming thinking deltas come as input_json_delta inside ThinkingBlock —
        # we can't easily tell them apart from tool-input deltas mid-stream, so
        # we leave thinking surfacing to the final-message pass below.

    def _final_to_turn(self, final: Any) -> ChatTurn:
        """Convert SDK final Message → our ChatTurn."""
        blocks_raw = getattr(final, "content", []) or []
        # Serialize each block to a dict so we can echo it back verbatim
        assistant_blocks: list[dict] = []
        text_parts: list[str] = []
        tool_calls: list[ToolCall] = []

        for b in blocks_raw:
            btype = getattr(b, "type", "")
            if btype == "text":
                t = getattr(b, "text", "")
                assistant_blocks.append({"type": "text", "text": t})
                if t:
                    text_parts.append(t)
            elif btype == "thinking":
                # Echo thinking blocks back unchanged so "extended thinking"
                # continuity holds across turns when supported.
                assistant_blocks.append({
                    "type": "thinking",
                    "thinking": getattr(b, "thinking", ""),
                    **({"signature": b.signature} if getattr(b, "signature", None) else {}),
                })
            elif btype == "redacted_thinking":
                assistant_blocks.append({
                    "type": "redacted_thinking",
                    "data": getattr(b, "data", ""),
                })
            elif btype == "tool_use":
                tc = ToolCall(
                    id=getattr(b, "id", "") or "",
                    name=getattr(b, "name", "") or "",
                    args=getattr(b, "input", {}) or {},
                )
                tool_calls.append(tc)
                assistant_blocks.append({
                    "type": "tool_use",
                    "id": tc.id,
                    "name": tc.name,
                    "input": tc.args,
                })
            else:
                # Unknown block type — try a best-effort serialization
                try:
                    assistant_blocks.append(json.loads(b.model_dump_json()))  # pydantic
                except Exception:
                    assistant_blocks.append({"type": btype or "unknown"})

        usage_obj = getattr(final, "usage", None)
        usage = Usage(
            in_tokens=getattr(usage_obj, "input_tokens", 0) or 0,
            out_tokens=getattr(usage_obj, "output_tokens", 0) or 0,
        )

        return ChatTurn(
            assistant_blocks=assistant_blocks,
            assistant_text="".join(text_parts),
            tool_calls=tool_calls,
            usage=usage,
            stop_reason=getattr(final, "stop_reason", "") or "",
        )

File: internal/agent/providers/anthropic.go
// Anthropic provider — wraps github.com/anthropics/anthropic-sdk-go.
//
// Handles three sources behind the same code path:
//
//	BaseURL=""                              → Anthropic direct
//	BaseURL="https://opencode.ai/zen/go"    → opencode-go
//	BaseURL=<custom>                        → any Anthropic-compatible gateway
//
// Streaming via client.Messages.NewStreaming. Per-token text deltas are
// forwarded to events.Delta. Thinking blocks are surfaced via
// events.Thinking and echoed verbatim on the next turn so extended-
// thinking continuity holds.
//
// Tool messages from our internal shape (role="tool") are repacked into
// Anthropic's user-role tool_result blocks before sending.
package providers

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/Flameingmoy/automedal/internal/agent"
	anthropic "github.com/anthropics/anthropic-sdk-go"
	anthropicopt "github.com/anthropics/anthropic-sdk-go/option"
)

// AnthropicProvider talks to anthropic.com or any Anthropic-compatible
// endpoint via BaseURL.
type AnthropicProvider struct {
	ModelName string
	APIKey    string
	BaseURL   string
	MaxTokens int64
	Timeout   time.Duration
}

// Model satisfies ChatProvider.
func (p *AnthropicProvider) Model() string { return p.ModelName }

func (p *AnthropicProvider) client() anthropic.Client {
	opts := []anthropicopt.RequestOption{anthropicopt.WithAPIKey(p.APIKey)}
	if p.BaseURL != "" {
		opts = append(opts, anthropicopt.WithBaseURL(p.BaseURL))
	}
	if p.Timeout > 0 {
		opts = append(opts, anthropicopt.WithRequestTimeout(p.Timeout))
	}
	return anthropic.NewClient(opts...)
}

func (p *AnthropicProvider) maxTokens() int64 {
	if p.MaxTokens > 0 {
		return p.MaxTokens
	}
	return 4096
}

// ChatStream sends one turn through the Anthropic streaming endpoint.
func (p *AnthropicProvider) ChatStream(ctx context.Context, in ChatRequest) (*ChatTurn, error) {
	c := p.client()
	apiMsgs, err := toAnthropicMessages(in.Messages)
	if err != nil {
		return nil, err
	}
	params := anthropic.MessageNewParams{
		Model:     anthropic.Model(p.ModelName),
		MaxTokens: p.maxTokens(),
		Messages:  apiMsgs,
	}
	if in.System != "" {
		params.System = []anthropic.TextBlockParam{{Text: in.System}}
	}
	if len(in.Tools) > 0 {
		params.Tools = toAnthropicTools(in.Tools)
	}

	final, err := agent.WithRetry(ctx, func() (anthropic.Message, error) {
		return p.runStream(ctx, &c, params, in.Events)
	}, agent.RetryOpts{
		Sink:  in.Events,
		Label: "anthropic.chat_stream model=" + p.ModelName,
	})
	if err != nil {
		return nil, err
	}
	return finalToChatTurn(&final, in.Events), nil
}

// runStream executes one streaming attempt + accumulates the events into
// a final Message via the SDK's Accumulate helper. Forwards text + thinking
// deltas to the EventSink as they arrive.
func (p *AnthropicProvider) runStream(
	ctx context.Context,
	c *anthropic.Client,
	params anthropic.MessageNewParams,
	events *agent.EventSink,
) (anthropic.Message, error) {
	stream := c.Messages.NewStreaming(ctx, params)
	defer stream.Close()

	var msg anthropic.Message
	for stream.Next() {
		ev := stream.Current()
		if err := msg.Accumulate(ev); err != nil {
			return anthropic.Message{}, err
		}
		// Forward text deltas live to the EventSink.
		if events == nil {
			continue
		}
		if cb, ok := ev.AsAny().(anthropic.ContentBlockDeltaEvent); ok {
			switch d := cb.Delta.AsAny().(type) {
			case anthropic.TextDelta:
				if d.Text != "" {
					events.Delta(d.Text)
				}
			case anthropic.ThinkingDelta:
				if d.Thinking != "" {
					events.Thinking(d.Thinking)
				}
			}
		}
	}
	if err := stream.Err(); err != nil {
		return anthropic.Message{}, err
	}
	return msg, nil
}

// ── outbound message translation ────────────────────────────────────────

func toAnthropicMessages(internal []agent.Message) ([]anthropic.MessageParam, error) {
	out := []anthropic.MessageParam{}
	pendingResults := []anthropic.ContentBlockParamUnion{}

	flushTools := func() {
		if len(pendingResults) > 0 {
			out = append(out, anthropic.NewUserMessage(pendingResults...))
			pendingResults = nil
		}
	}

	for _, m := range internal {
		switch m.Role {
		case "tool":
			content, _ := m.Content.(string)
			pendingResults = append(pendingResults, anthropic.NewToolResultBlock(m.ToolUseID, content, m.IsError))
			continue
		}
		flushTools()

		switch m.Role {
		case "user":
			switch c := m.Content.(type) {
			case string:
				out = append(out, anthropic.NewUserMessage(anthropic.NewTextBlock(c)))
			case []agent.Block:
				blocks, err := blocksToParams(c)
				if err != nil {
					return nil, err
				}
				out = append(out, anthropic.NewUserMessage(blocks...))
			default:
				return nil, fmt.Errorf("unexpected user content type %T", m.Content)
			}
		case "assistant":
			blocks, err := blocksToParams(m.Content.([]agent.Block))
			if err != nil {
				return nil, err
			}
			out = append(out, anthropic.NewAssistantMessage(blocks...))
		default:
			return nil, fmt.Errorf("unknown role %q", m.Role)
		}
	}
	flushTools()
	return out, nil
}

func blocksToParams(blocks []agent.Block) ([]anthropic.ContentBlockParamUnion, error) {
	out := make([]anthropic.ContentBlockParamUnion, 0, len(blocks))
	for _, b := range blocks {
		switch b.Type {
		case "text":
			out = append(out, anthropic.NewTextBlock(b.Text))
		case "thinking":
			out = append(out, anthropic.NewThinkingBlock(b.Signature, b.Thinking))
		case "redacted_thinking":
			out = append(out, anthropic.NewRedactedThinkingBlock(b.Data))
		case "tool_use":
			out = append(out, anthropic.NewToolUseBlock(b.ID, b.Input, b.Name))
		case "tool_result":
			s, _ := b.Content.(string)
			out = append(out, anthropic.NewToolResultBlock(b.ToolUseID, s, b.IsError))
		default:
			return nil, fmt.Errorf("unknown block type %q", b.Type)
		}
	}
	return out, nil
}

func toAnthropicTools(specs []ToolSpec) []anthropic.ToolUnionParam {
	out := make([]anthropic.ToolUnionParam, 0, len(specs))
	for _, s := range specs {
		props, _ := s.Schema["properties"]
		req, _ := s.Schema["required"].([]string)
		if req == nil {
			if rawReq, ok := s.Schema["required"].([]any); ok {
				req = make([]string, len(rawReq))
				for i, v := range rawReq {
					req[i], _ = v.(string)
				}
			}
		}
		tp := anthropic.ToolParam{
			Name: s.Name,
			InputSchema: anthropic.ToolInputSchemaParam{
				Properties: props,
				Required:   req,
			},
		}
		if s.Description != "" {
			tp.Description = anthropic.String(s.Description)
		}
		out = append(out, anthropic.ToolUnionParam{OfTool: &tp})
	}
	return out
}

// ── inbound (final → ChatTurn) ──────────────────────────────────────────

func finalToChatTurn(m *anthropic.Message, _ *agent.EventSink) *ChatTurn {
	turn := &ChatTurn{
		Usage: agent.Usage{
			InTokens:  int(m.Usage.InputTokens),
			OutTokens: int(m.Usage.OutputTokens),
		},
		StopReason: string(m.StopReason),
	}
	for _, cb := range m.Content {
		switch cb.Type {
		case "text":
			turn.AssistantBlocks = append(turn.AssistantBlocks, agent.Block{
				Type: "text",
				Text: cb.Text,
			})
			turn.AssistantText += cb.Text
		case "thinking":
			turn.AssistantBlocks = append(turn.AssistantBlocks, agent.Block{
				Type:      "thinking",
				Thinking:  cb.Thinking,
				Signature: cb.Signature,
			})
		case "redacted_thinking":
			turn.AssistantBlocks = append(turn.AssistantBlocks, agent.Block{
				Type: "redacted_thinking",
				Data: cb.Data,
			})
		case "tool_use":
			var input map[string]any
			if len(cb.Input) > 0 {
				_ = json.Unmarshal(cb.Input, &input)
			}
			turn.AssistantBlocks = append(turn.AssistantBlocks, agent.Block{
				Type:  "tool_use",
				ID:    cb.ID,
				Name:  cb.Name,
				Input: input,
			})
			turn.ToolCalls = append(turn.ToolCalls, ToolCall{
				ID:   cb.ID,
				Name: cb.Name,
				Args: input,
			})
		}
	}
	return turn
}
